### PR TITLE
doc: add GNOME

### LINKED ISSUE #16285: Document wrapGAppsHook
I just found out about wrapGAppsHook by examining a gnome package, but there really should be documentation about this in the Nixpkgs User Guide (a section about Gnome would be nice, which would show the right way to package GTK3 / GTK2 glib based apps).

Also, perhaps there are other things I missed that one should know when packaging Gnome Applications or GTK3/2 apps? 

/cc @lethalman @jgeerds @DamienCassou Gnome maintainers

### PR DESCRIPTION
**Pull request by jtojnar**
Work in progress documentation on packaging GNOME ecosystem applications. Comments and questions welcome.

Closes: #16285

### CODE REVIEW

File: doc/languages-frameworks/gnome.xml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        Some applications will also require <package>gsettings-desktop-schemas</package> for things like reading proxy configuration or user interface customization. This dependency is often not mentioned by upstream, you should grep for <literal>org.gnome.desktop</literal> and <literal>org.gnome.system</literal> to see if the schemas are needed.
+      </para>
+    </section>
+
+
+    <section xml:id="ssec-gnome-icons">
+      <title>Icons</title>
+
+      <para>
+        When an application uses icons, an icon theme should be available in <envar>XDG_DATA_DIRS</envar>. The default <link xlink:href="https://www.freedesktop.org/wiki/Software/icon-theme/">hicolor-icon-theme</link> is recommended, unless the application needs some specific icons not available in there.
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]: Apparently, `hicolor-icon-theme` does not contain any icons. Nevertheless, we should mention it for its setup hook that removes `index.theme` files.

File: doc/languages-frameworks/gnome.xml [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+          </para>
+        </listitem>
+        <listitem>
+          <para>
+            <package>hicolor-icon-theme</package>’s setup hook will add themes to <envar>XDG_ICON_DIRS</envar> which is appended to <envar>XDG_DATA_DIRS</envar> by <package>wrapGAppsHook</package>.
+          </para>
+        </listitem>
+        <listitem>
+          <para>
+            <package>gobjectIntrospection</package> setup hook populates <envar>GI_TYPELIB_PATH</envar> variable with <filename>lib/girepository-1.0</filename> directories of dependencies, which is then added to wrapper by <package>wrapGAppsHook</package>. It also adds <filename>share</filename> directories of dependencies to <envar>XDG_DATA_DIRS</envar>, which is intended to promote GIR files but it also <link xlink:href="https://github.com/NixOS/nixpkgs/issues/32790">pollutes the closures</link> of packages using <package>wrapGAppsHook</package>.
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]: We need to fix the pollution.

**Comment by jtojnar** [comment]
Regarding GStreamer this is useful: https://gstreamer.freedesktop.org/documentation/plugins.html

**Comment by cdepillabout** [comment]
This documentation was helpful in getting icons to actually work in a [GTK app](https://github.com/cdepillabout/termonad/pull/91) I am developing.

It would be really nice to get this in the nixpkgs manual.

The one problem I did run into was with `hicolor-icon-theme`, as mentioned above.  The documentation makes it sound like all you need to do is add this as a `buildInput` and icons should be visible.  However, it doesn't look like the `hicolor-icon-theme` derivation actually contains any icons.  Instead I had to use `gnome3.adwaita-icon-theme`.

Other than that, this looks good and is a :+1: from me.

**Comment by jtojnar** [comment]
> The one problem I did run into was with `hicolor-icon-theme`, as mentioned above. The documentation makes it sound like all you need to do is add this as a `buildInput` and icons should be visible. However, it doesn't look like the `hicolor-icon-theme` derivation actually contains any icons. Instead I had to use `gnome3.adwaita-icon-theme`.

Yup, figured that out as well. Actually, nowadays I just prefer to let users install an icon theme of their own choosing.

**Comment by jtojnar** [comment]
Unfortunately, there is no other way – you either include an icon theme in the package’s closure, or you might not have icons on insufficiently configured systems. (Side note: It would be nice to have a [mechanism for specifying system dependencies](https://github.com/NixOS/nixpkgs/issues/52777#issuecomment-449866799) but we are not there yet.)

In my opinion, if you do not use DE, you are expected to create it yourself.

```
xdg.icons.enable = true;
environment.systemPackages = [ pkgs.gnome3.adwaita-icon-theme ];
```

should be enough.

I really need to write a chapter on creating a baseline freedesktop system though, as many NixOS users run some kind of bare-bones WM.

**Comment by cdepillabout** [comment]
> I really need to write a chapter on creating a baseline freedesktop system though, as many NixOS users run some kind of bare-bones WM.

Yeah, this would be really nice.  I run XMonad, but not configured from `/etc/nixos/configuration.nix`.  It would be great to have a guide as to how all the freedesktop stuff is supposed to work in nix.

**Comment by jtojnar** [comment]
> What about people e.g. using nixpkgs on a non-NixOS distro?

They can still install the icon theme, either with nix or their native package manager.

> I think that such issues that can be solved at package-level should be solved at package-level to avoid delivering completely broken applications on non-GNOME environments.

Well, the main issue with depending on `adwaita-icon-theme` is that it increases closure size for everyone, even when want to use a different icon set. Additionally, it depends on `librsvg` for build; though this is not currently a problem, as `librsvg` is already needed for `wrapGAppsHook`.

**Comment by worldofpeace** [comment]
Don't fret, just resolved the conflict and added a little commit separately.

**Comment by jtojnar** [comment]
Here I am hinting on the future of GLib infrastructure in nixpkgs:

https://github.com/NixOS/nixpkgs/pull/67320#issuecomment-524158181

File: doc/languages-frameworks/gnome.xml [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    backend. You will need to add <literal>dconf</literal> GIO module to
+    <envar>GIO_EXTRA_MODULES</envar> variable, otherwise the
+    <literal>memory</literal> backend will be used and the saved settings will
+    not be persistent.
+   </para>
+
+   <para>
+    Last you will need the dconf database itself. Installing
+    <package>gnome3.dconf</package> to
+    <option>environment.systemPackages</option> will be enough.
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]: Change this to nixos option.

File: doc/languages-frameworks/gnome.xml [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+  <section xml:id="ssec-gnome-icons">
+   <title>Icons</title>
+
+   <para>
+    When an application uses icons, an icon theme should be available in
+    <envar>XDG_DATA_DIRS</envar>. The default
+    <link xlink:href="https://www.freedesktop.org/wiki/Software/icon-theme/">hicolor-icon-theme</link>
+    is recommended, unless the application needs some specific icons not
+    available in there.
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]: Yeah, this is nonsense. Anyone know why are we adding this other than for clean-up hook?

File: doc/languages-frameworks/gnome.xml [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+     <para>
+      <package>glib</package> setup hook will populate
+      <envar>GSETTINGS_SCHEMAS_PATH</envar> and then
+      <package>wrapGAppsHook</package> will append it to
+      <envar>XDG_DATA_DIR</envar>.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      <package>gnome3.dconf.lib</package> is a dependency of
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]: Should we move this out of `gnome3`?

File: doc/languages-frameworks/gnome.xml [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+     <para>
+      <package>wrapGAppsHook</package> itself will add the package’s
+      <filename>share</filename> directory to <envar>XDG_DATA_DIR</envar>.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      <package>glib</package> setup hook will populate
+      <envar>GSETTINGS_SCHEMAS_PATH</envar> and then
+      <package>wrapGAppsHook</package> will append it to
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]:
```suggestion
      <package>wrapGAppsHook</package> will prepend it to
```

File: doc/languages-frameworks/gnome.xml [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    <listitem>
+     <para>
+      <package>gnome3.dconf.lib</package> is a dependency of
+      <package>wrapGAppsHook</package>, which then also adds it to the
+      <envar>GIO_EXTRA_MODULES</envar> variable.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      <package>hicolor-icon-theme</package>’s setup hook will add themes to
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]:
```suggestion
      <package>hicolor-icon-theme</package>’s setup hook will add icon themes to
```

File: doc/languages-frameworks/gnome.xml [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+      <envar>GRL_PLUGIN_PATH</envar> variables, respectively, which will then
+      be added to the wrapper by <literal>wrapGAppsHook</literal>.
+     </para>
+    </listitem>
+   </itemizedlist>
+  </para>
+
+  <para>
+   You can also pass additional arguments to <literal>makeWrapper</literal>
+   using <literal>gappsWrapperArgs</literal>:
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]: Emphasize that this should happen in `preFixup`.

File: doc/languages-frameworks/gnome.xml [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    <listitem>
+     <para>
+      <package>gobject-introspection</package> setup hook populates
+      <envar>GI_TYPELIB_PATH</envar> variable with
+      <filename>lib/girepository-1.0</filename> directories of dependencies,
+      which is then added to wrapper by <package>wrapGAppsHook</package>. It
+      also adds <filename>share</filename> directories of dependencies to
+      <envar>XDG_DATA_DIRS</envar>, which is intended to promote GIR files but
+      it also
+      <link xlink:href="https://github.com/NixOS/nixpkgs/issues/32790">pollutes
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]: Still not sure how serious this is.

File: doc/languages-frameworks/gnome.xml [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        '';
+      </programlisting>
+  </para>
+ </section>
+
+ <section xml:id="ssec-gnome-updating">
+  <title>Updating GNOME packages</title>
+
+  <para>
+   Most GNOME package offer <literal>updateScript</literal>, it is therefore
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]: link to updateScript docs.

File: doc/languages-frameworks/gnome.xml (hunk 4 above)
Comment (worldofpeace) [reply doc/languages-frameworks/gnome.xml]: It seems people add it when they're running under systems missing icon themes, and they see
* https://gitlab.gnome.org/GNOME/gtk/blob/master/gtk/gtkicontheme.c#L1861

Comment (worldofpeace) [reply doc/languages-frameworks/gnome.xml]: And I guess XDG_ICON_DIRS

File: doc/languages-frameworks/gnome.xml [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+      <package>wrapGAppsHook</package> itself will add the package’s
+      <filename>share</filename> directory to <envar>XDG_DATA_DIR</envar>.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      <package>glib</package> setup hook will populate
+      <envar>GSETTINGS_SCHEMAS_PATH</envar> and then
+      <package>wrapGAppsHook</package> will append it to
+      <envar>XDG_DATA_DIR</envar>.
```
Comment (worldofpeace) [inline doc/languages-frameworks/gnome.xml]:
```suggestion
      <envar>XDG_DATA_DIRS</envar>.
```

File: doc/languages-frameworks/gnome.xml [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+     <para>
+      <package>gnome3.dconf.lib</package> is a dependency of
+      <package>wrapGAppsHook</package>, which then also adds it to the
+      <envar>GIO_EXTRA_MODULES</envar> variable.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      <package>hicolor-icon-theme</package>’s setup hook will add themes to
+      <envar>XDG_ICON_DIRS</envar> which is appended to
```
Comment (worldofpeace) [inline doc/languages-frameworks/gnome.xml]: Is this also prepended

**Comment by jtojnar** [comment]
Updated with issues fixed and some new content.

Rendered: https://jtojnar.github.io/dumpling/#sec-language-gnome

**Review by worldofpeace — CHANGES_REQUESTED** [review]
##### General thoughts and notes while reading the sections

* `GTK Themes`
maybe not say it's no longer needed. I know for Pantheon their theme is needed or it's all broken, and they're still gtk apps.

* `Various plug-ins`
I think we have duplicated docs on gstreamer, this one https://github.com/NixOS/nixpkgs/blob/ac2eb3151df879602c58280d09e9c205fcff8534/doc/stdenv.xml#L2636 explaning the GST_PLUGIN_SYSTEM_PATH variable.

*  `Icons`
Does gtk require adwaita-icon-theme, at least as a fallback on the system?
If so, we could say to install it.

---

I had some thoughts on additions / clarifications, I think I'll push them tomorrow.

**Comment by ryantm** [comment]
@jtojnar I read through the section and didn't notice any grammatical or typographical errors. I don't have the background to verify the technical aspects, but it sounded reasonable.

**Comment by jtojnar** [comment]
>     * `GTK Themes`
>       maybe not say it's no longer needed. I know for Pantheon their theme is needed or it's all broken, and they're still gtk apps.

:heavy_check_mark: Fixed.

>     * `Various plug-ins`
>       I think we have duplicated docs on gstreamer, this one https://github.com/NixOS/nixpkgs/blob/ac2eb3151df879602c58280d09e9c205fcff8534/doc/stdenv.xml#L2636
>        explaning the GST_PLUGIN_SYSTEM_PATH variable.

Hmm, should we move the whole setup hook section to stdenv? On the other hand, this locality is useful when reading the GNOME subchapter as a whole.

>     * `Icons`
>       Does gtk require adwaita-icon-theme, at least as a fallback on the system?
>       If so, we could say to install it.
> 

From my test with Nautilus with `adwaita-icon-theme` carved out of the expression, GTK will always use [`gtk-icon-theme-name`](https://developer.gnome.org/gtk3/unstable/GtkSettings.html#GtkSettings--gtk-icon-theme-name) setting and fallback to `hicolor` for missing icons (e.g. application icons). So with the setting being `elementary`, it will not touch Adwaita icon theme. We already mention that user needs to install the theme but these details are better suited to a “Creating a custom DE” in NixOS manual.

**Comment by worldofpeace** [comment]
> Hmm, should we move the whole setup hook section to stdenv? On the other hand, this locality is useful when reading the GNOME subchapter as a whole.

Perhaps we should remove it from stdenv, and just keep it in GNOME subchapter. I'd like those setup hooks to be documented in the locality of the GNOME sub-chapter.

**Comment by jtojnar** [comment]
Updated the stdenv section:

image

**Comment by worldofpeace** [comment]
@jtojnar That's perfect :+1:

File: doc/languages-frameworks/gnome.xml [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+
+  <para>
+   You can also pass additional arguments to <literal>makeWrapper</literal>
+   using <literal>gappsWrapperArgs</literal> in <literal>preFixup</literal>
+   hook:
+<programlisting>
+preFixup = ''
+  gappsWrapperArgs+=(
+    # Thumbnailers
+    --prefix XDG_DATA_DIRS : "${gdk_pixbuf}/share"
```
Comment (jtojnar) [inline doc/languages-frameworks/gnome.xml]:
```suggestion
    --prefix XDG_DATA_DIRS : "${gdk-pixbuf}/share"
```

**Comment by jtojnar** [comment]
Backported in 9d8e16173d0

**Pull request merged**